src/components/App/App.jsx

```diff
@@ -0,0 +1,283 @@
+/*   PLEASE READ THE README FILE */
+
+import { Route, Routes, useNavigate, useLocation } from "react-router-dom";
+import { useState, useEffect } from "react";
+import "./App.css";
+import Header from "../Header/Header";
+import Main from "../Main/Main";
+import Footer from "../Footer/Footer";
+import RegisterModal from "../RegisterModal/RegisterModal";
+import LoginModal from "../LoginModal/LoginModal";
+import EditProfileModal from "../EditProfileModal/EditProfileModal";
+import RecipeModal from "../RecipeModal/RecipeModal";
+import SearchedRecipes from "../SearchedRecipes/SearchedRecipes";
+import Profile from "../Profile/Profile";
+import ProtectedRoute from "../ProtectedRoute/ProtectedRoute";
+import About from "../About/About";
+import CurrentUserContext from "../../contexts/CurrentUserContext";
+import { getToken, setToken } from "../../utils/token";
+import {
+  registerUser,
+  signInUser,
+  updateUser,
+  isValidToken,
+} from "../../utils/auth";
+import { deleteRecipeCard, saveRecipe, getRecipeItems } from "../../utils/api";
+
+function App() {
+  const [activeModal, setActiveModal] = useState("");
+  //const [recipes, setRecipes] = useState([]);
+  const [savedRecipes, setSavedRecipes] = useState([]);
+  const [recommended, setRecommended] = useState([]);
+  const [currentUser, setCurrentUser] = useState(null);
+  const [isLoggedIn, setIsloggedIn] = useState(false);
+  const [selectedRecipe, setSelectedRecipe] = useState(null);
+
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  const handleRegisterUser = () => {
+    setActiveModal("register");
+  };
+
+  const handleLoginUser = () => {
+    setActiveModal("login");
+  };
+
+  const handleEditProfileUser = () => {
+    setActiveModal("edit-profile");
+  };
+
+  const handleRecipeSummaryOpen = (recipe) => {
+    setActiveModal("summary");
+    setSelectedRecipe(recipe);
+    console.log(recipe);
+  };
+
+  const closeActiveModal = () => {
+    setActiveModal("");
+  };
+
+  const handleRegistration = ({ email, password, name }) => {
+    registerUser({
+      email,
+      password,
+      name,
+    })
+      .then((res) => {
+        setIsloggedIn(true);
+        setCurrentUser(res.data);
+        console.log(res);
+        navigate("/profile");
+        closeActiveModal();
+      })
+      .catch((res) => {
+        console.error(res);
+      });
+  };
+
+  const handleLogin = ({ email, password }) => {
+    if (!email || !password) {
+      return;
+    }
+
+    signInUser({ email, password })
+      .then((res) => {
+        setToken(res.token);
+        return isValidToken(res.token);
+      })
+      .then((user) => {
+        setCurrentUser(user);
+        setIsloggedIn(true);
+        closeActiveModal();
+        navigate("/");
+      })
+      .catch((err) => {
+        console.error("Wrong Login information", err);
+      });
+  };
+
+  const handleEditProfile = ({ name }) => {
+    updateUser({ name })
```

Comment:
**COULD BE IMPROVED**

You can create a special variable for watching **loading** states during server requests:

```js
 const [isLoading, setIsLoading] = React.useState(false);

```

And then you can change it to `true` before sending requests and return it back to `false`  in `finally` to stop loading:

And you should pass `isLoading` to every popup to change the **button text**:

```js

 buttonText={isLoading? 'Saving...' : 'Save'}

```